Detect cgroups and namespaces via virt-launcher-socket

rmohr opened the pull request:
Instead of introspecting docker containers, to find out the Pod namespaces and cgroups, let virt-handler detect them via a socket. Every virt-launcher instance creates a socket in a directory which is shared with virt-handler. On the VM start, virt-handler checks which process owns this socket and determines which cgroups and namespaces to use for the qemu process. This PR does not yet treat the socket creation as a trigger for the virt-handler controller. This will come in another PR. 

Further, the PR does not use the `<resource><partition></partition></resource>` field in the Domain XML, to set the cgroups. This turns out to be more complex than expected. If libvirt does not have access to systemd, it will try to create it's own cgroup hierarchy and will therefore add `.partition`, where it should add `.slice`. If it has access to systemd, it will try to add `.slice` as group suffix, where k8s/systemd uses `.scope`.

berrange commented on pkg/virt-handler/virtwrap/isolation/isolation.go:
```diff
@@ -0,0 +1,129 @@
+		if slice == "" {
+			slice = cgEntry[2]
+		} else if slice != cgEntry[2] {
+			err = fmt.Errorf("Process is part of more than one slice. Expected %s, found %s", slice, cgEntry[2])
```
This is fragile - it would not be unusual to see a new cgroup present and it not be used by docker, or for use of some cgroup controllers to have been disabled. Libvirt only deals with a subset of controllers, so we should just validate the controllers that we actually need.

rmohr replied:
Will do that.

berrange commented on pkg/virt-handler/virtwrap/isolation/isolation.go:
```diff
@@ -0,0 +1,129 @@
+	ufile, err := sock.(*net.UnixConn).File()
+	if err != nil {
+		return -1, err
+	}
```
You're letting ufile go out of scope which means the socket will get closed. This creates a race, because once the socket is closed, the launcher pod could go away & be replaced by another process making the PID credentials you get invalid. We need to keep the socket open for as long as we're managing the process.

fabiand replied:
The socket will be kept open for longer once we do more with it.

rmohr commented on pkg/virt-handler/virtwrap/isolation/isolation.go:
```diff
@@ -0,0 +1,129 @@
+	ufile, err := sock.(*net.UnixConn).File()
+	if err != nil {
+		return -1, err
+	}
```
I don't think that keeping the FileDescriptor open helps avoiding that race. We would still have already created and started the VM, before we would detect that the file descriptor is now closed. Also see https://github.com/kubevirt/kubevirt/pull/337#issuecomment-316970884 how this fits into the whole picture.

rmohr commented:
@berrange  thx for the review. 

So regarding the socket creation flow change:

The final story, which would be completed by the next PR, which will use the sockets also as source for the controller inside virt-handler is this:

virt-launchers will create the sockets in a flat directory structure. virt-handler watches via a ListWatcher, based on listing all sockets in the directory (list) and inotify (watch) in the directory for creations and deletions of sockets. This way, we can follow in the controller the creations and deletions. Like with a file descriptor, which would be kept open the whole time, we can now follow these socket changes. In both ways there would be potentially the risk of a race. Both ways, by themselves alone don't provide enough guarantee. Right now, the guarantee, that the VM never gets started or dies immediately when the pod changes, does the PID namespace change.

Advange of this approach, compared to connecting to one socket maintained by virt-handler:

 * No need keep the file descriptors open
   + This means, we can resync and restart virt-handler anytime
   + virt-launcher is independent of virt-handler. It does not need any reconnect logic, or anything else ...
 * A ListWatcher source much better suits to the watch and follow approach of the k8s client-go framework
 * We still have the same race guarantees like with the other approach

berrange commented:
You can't rely on inotify to track deletion of UNIX socket files on disk. Deleting the UNIX socket requires explicit action during shutdown, and so that will fail if the process is terminated abnormally. You must have the UNIX sockets open to detect EOF as that's guaranteed to occur if the process has gone. Having multiple different unprivileged process all create sockets in the same directory is also a bad idea - its the classic scenario that's lead to countless /tmp file vulnerabilities.

stu-gott commented:
Both approaches have their trade-offs. One the one hand you have the possibility that a virt-launcher process dies abnormally and leaves a dangling socket file. And on the other you have the possibility that virt-handler dies abnormally and does the same.

@berrange In the handler-opens-socket scenario, how should virt-launcher react if the connection to virt-handler is broken (e.g. during a normal re-start)?

rmohr commented:
> You can't rely on inotify to track deletion of UNIX socket files on disk. Deleting the UNIX socket requires explicit action during shutdown, and so that will fail if the process is terminated abnormally. You must have the UNIX sockets open to detect EOF as that's guaranteed to occur if the process has gone. 

That is a good point. I will have think of something else then. If I can't restart virt-handler and need to have the connection open the whole time,  I am not very happy with that solution.

> Having multiple different unprivileged process all create sockets in the same directory is also a bad idea - its the classic scenario that's lead to countless /tmp file vulnerabilities.

We have to somehow verify the other end anyway. With a k8s secret or something, which we ask for over the socket ...

rmohr commented:
A quick and dirty solution, would of course be, to resync every 30 seconds, and also open the sockets from the virt-handler side, to see if there is still someone listening. Even one minute would be fine. It is not critical for us if we miss it, since we can't rely on the socket for VM start/dies anyway. It is just a nice addition to react in time (in that context, I am still not sure what our hard PID namespace guarantee should be replaced with).

davidvossel commented:
> This socket creation is the wrong way around. virt-handler should be creating a single listener socket, that is then connected to by virt-launcher, which will trigger startup. With this inverted approach, you're forcing virt-handler to trigger startup via some other means, since I has no idea when the socket has been created.

maybe help clarify my understanding here. I might be missing something.

I had it in my mind that virt-handler was going to create a listener socket like you've described, and that virt-launcher was going to create a unix socket as well like @rmohr's patch does now.  Then when virt-launcher initiates it would connect to virt-handler which would both tell virt-handler about about the unix socket virt-launcher created and trigger VM launch. From there, I thought there was some libvirt feature we were going to leverage where we hand off the file virt-launcher created to libvirt.

davidvossel commented:
Looking through these comments some more, this is shaping up to be something different than I expected.

We need to avoid any persistent connection between virt-launcher and virt-handler/libvirtd that is coupled with the lifecycle of a vm.  Virt-handler and the libvirtd container must be able to fail without impacting active VMs on the host. We also need to maintain the ability to update all of these components independently of one another.

The reason I was okay with the socket approach is because I was led to believe virt-launcher would open a socket that would be associated with the qemu process. So, when the qemu process exits, we'd get an EOF in the corresponding virt-launcher container. If this isn't the case, then the socket approach is arguably worse than what we already have. 

@berrange just to clarify, does the feature to associate a socket with the lifecycle of a qemu process exist now, or was this something that we could theoretically do in the future?

rmohr commented:
In general, we could move forward with this, although it does not solve the pods-vm binding, and not the cgroup joining, it at least makes kubevirt pretty much CRI independent and provides an unambiguous way of detecting the right cgroups and namespaces. virt-handler, can also in a follow on PR check for a k8s secret to be present within the virt-launcher pod, which created the socket. 

Thoughts?

davidvossel commented:
> In general, we could move forward with this, although it does not solve the pods-vm binding, and not the cgroup joining, it at least makes kubevirt pretty much CRI independent and provides an unambiguous way of detecting the right cgroups and namespaces. virt-handler, can also in a follow on PR check for a k8s secret to be present within the virt-launcher pod, which created the socket.

@rmohr I agree.  This is way cleaner way of doing the cgroup and namespace detection than what we are currently doing.  I'll try and help get this merged by doing a detailed code review soon.

davidvossel commented on pkg/virt-handler/virtwrap/manager.go:
```diff
@@ -330,12 +337,29 @@ func VMNamespaceKeyFunc(vm *v1.VM) string {
 		return nil, errors.NewAggregate(mappingErrs)
 	}
 
+	res, err := l.podIsolationDetector.Detect(vm)
```
I think there might be an issue here.

It's good that we re-detect the cgroups/ns/pid every time we sync the vm, however I think there's an issue here if the virt-launcher fails and restarts.

We'll detect all the new cgroup/ns info after virt-launcher recovers, but if the domain was defined with the old info we never actually update the domain xml file.

We update wantedSpec, but wantedSpec only gets set as domainXML once and is never updated after that.

rmohr replied:
You are right

davidvossel commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -53,7 +53,11 @@ var _ = Describe("Template", func() {
+					"-qemu-timeout", "60s",
+					"-name", "testvm",
+					"-namespace", "testns",
+					"-socket-dir", "/var/run/libvirt"}))
```
In the normal runtime, socket-dir is always set to /var/run/kubevirt, but for some reason all the tests set it to /var/run/libvirt.  Is this intentional?

davidvossel commented on manifests/virt-handler.yaml.in:
```diff
@@ -29,6 +30,8 @@ spec:
         volumeMounts:
         - name: libvirt-runtime
           mountPath: /var/run/libvirt
+        - name: sockets
```
It's possible in the future that there will be more data related to a domain that needs to be shared between local KubeVirt components than just the unix file.  For example, the cloud-init nocloud data source is going to require us sharing a .iso file between virt-handler and libvirt. 

maybe we should consider making this a generic share folder instead of labeling it as "sockets".  Maybe something like "domain-share" or something like that?

davidvossel replied:
yeah, I agree, lets not derail this pr with a discussion about a generic shared directory.   We can pick up talking about this another time. IMO, The current approach in the pr is fine for now.

stu-gott commented on manifests/virt-handler.yaml.in:
```diff
@@ -29,6 +30,8 @@ spec:
         volumeMounts:
         - name: libvirt-runtime
           mountPath: /var/run/libvirt
+        - name: sockets
```
I'd prefer not to mix things like sockets and .iso files. Nothing inherently is wrong with that at face value but it might lead to corner cases that could have security implications.

davidvossel replied:
I was thinking each type data would be in a separate sub directory.

**share directory**
/var/run/kubevirt

**top level sub directory for all ephemeral data associated with a VM called 'myvm' in namespace 'default**
/var/run/kubevirt/default/myvm

**'myvm's socket subdirectory**
/var/run/kubevirt/default/myvm/sockets/sock

**'myvm's cloud init subdirectory**
/var/run/kubevirt/default/myvm/cloudInit/noCloud.iso

virt-launcher associated with some VM called 'myvm' would only have access to the 'myvm' subdirectory. Virt-handler and libvirt would have access to the whole /var/run/kubevirt/ directory tree.

By having all ephemeral data associated with a VM tied to a single top level directory, it makes garbage collection a little easier. We wouldn't have to keep up with all the different components needing to share data between components.  We'd just know to delete a top level sub directory if the VM is no longer around. 

Also, as new features come around that need to share things between components related to a domain, all they have to do is drop their data into the VM's subdirectory. There would be some guarantees about how that data is cleaned up that everyone gets for free.

stu-gott replied:
Thanks for clarifying. I like that approach.

fabiand replied:
I'd like to separate the problem a little: We need cloud-init no question. but is the virt-handler creating it? Or could there be another component which is taking a cloud-init datasource and providing it as a volume to a VM? or - would it even make sense of libvirt is capable of generating noclpoud sources based on the data?

But this should probably be discussed in #340.

rmohr commented:
> To me the value of this PR is unclear.

Right now it gives us cgroup and  namespace detection in a CRI independent way.

> The discussion would really benefit from a design shedding light on the controversial issues and giving reasons why the one or other approach was chosen.
> 
> Until then it's difficult to tell if it's the right direction.

 If we would want to switch the socket creation later we would have almost exactly the same detection code + much more reconnect logic. So if we need it, we can easily extend this code to do the socket creation the other way.

I think that we really don't have a clear picture yet, how to solve setting the cgroups and how to tie the VM lifetime to the pod lifetime without the PID namespace. For that I agree that we need ideas and proposals. The PR is not addressing these issues, so would defering that fine for you?

davidvossel commented:
> To me the value of this PR is unclear. There are still quite a few open questions

> What are the implications (pros and cons) of keeping the sockets open or closing them
The scope of this PR has changed through the course of the review. 

Initially this was seen as a way to move towards a new method of binding the VM lifecycle to the virt-launcher pod through the use of a socket. however through further review (history is in this pr), It appears using the socket as a means of coordination is not a good option at this point. 

The remaining value here is we have a way of detecting cgroups and namespaces that is independent of Docker.  With cgroup/namespace detection as the only goal here, this PR seems fairly sane and useful to me.

The component that lives in the cgroup/namespace we want to detect (virt-launcher) creates the unix socket.  The other side (virt-handler) uses the socket to get pid info and ultimately cgroup/namespace info in a way that is not container runtime dependent.

With that scope in mind, There's no value in keeping the socket open. We just want open the socket in order to determine the virt-launcher pid.

> Handler vs launcher opening/providing sockets

If the socket is just for a runtime independent way of pid detection, it appears to make the most sense for virt-launcher to create the socket since that's the component we want to discover the cgroup/namespace 
of.

> the issue around resource tag

what is the issue around resource tag you're referring to?

fabiand commented on pkg/virt-handler/virtwrap/isolation/isolation.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func SocketFromNamespaceName(baseDir string, namespace string, name string) string {
+	return filepath.Clean(baseDir) + "/" + namespace + "/" + name + "/sock"
```
This is a security risk.

There are eventually few checks on what `name` can contain. An alterantive is to use the VM's UUID

fabiand replied:
Once you have a new way to create the path names, please document it somewhere.

A similar mechanism is actually described here: https://github.com/kubernetes/community/pull/589/files#diff-c1f7e2debf21346978bae50b5f7e37d2R91

fabiand commented on pkg/virt-handler/virtwrap/api/schema.go:
```diff
@@ -142,15 +142,31 @@ type DomainList struct {
+	SysInfo  *SysInfo     `xml:"sysinfo,omitempty"`
+	Devices  Devices      `xml:"devices"`
+	Clock    *Clock       `xml:"clock,omitempty"`
+	Resource *Resource    `xml:"resource,omitempty"`
```
Is there a bug or RFE filed against libvirt to improve the resource field so we can use it?

fabiand commented on pkg/virt-handler/virtwrap/api/schema.go:
```diff
@@ -142,15 +142,31 @@ type DomainList struct {
+	QEMUCmd  *Commandline `xml:"qemu:commandline,omitempty"`
+}
+
+type Commandline struct {
```
We should switch to libvirt-xml-go quickly - instead of writing our own mapping.

fabiand replied:
Or are there significant gaps?

rmohr replied:
Constructing the structs like I like them, and then converting them to the XML expectations for libvirt works fine for me. Fields which are not there, are definitely fields we don't support, I like this kind of white-listing ...

Feel free to switch over to libvirt-xml-go, I don't see drawbacks, but I don't see much benefit either.

fabiand replied:
The benefit is to avoid code duplication.

rmohr replied:
Sure, I will not stop you doing the switch.

rmohr commented:
> An RFE/bug must be filed on libvirt for the Resource field
RFE is now there: https://bugzilla.redhat.com/show_bug.cgi?id=1487336

rmohr commented:
Additional test and documentation added. Fixing the VM re-apply bug will follow tomorrow.

rmohr commented:
Ok, only the usual test fails. I implemented a non-final workaround, on how to share host specific data for migrations.

Before we merge this, I want to add one or two documents, to describe how the VM start and migration flow works, next refactor steps and where we want to go (as far as we know). However, please do a code-review again in the meantime.

davidvossel commented on cmd/virt-migrator/migrate:
```diff
@@ -22,23 +22,47 @@ case $key in
 
 # Tell libvirt where qemu will listen for spice connections on the new host
 virsh -c $SOURCE dumpxml $DOMAIN > $DOMAIN.xml
+cat $DOMAIN.xml
```
did you mean to leave this in the script, or was it just being used for debug?

davidvossel commented on pkg/virt-handler/virtwrap/manager.go:
```diff
@@ -203,19 +224,28 @@ func (l *LibvirtDomainManager) SyncVM(vm *v1.VM) (*api.DomainSpec, error) {
+	// To make sure, that we set the right qemu wrapper arguments,
+	// we update the domain XML whenever a VM was already defined but not running
+	if !newDomain && cli.IsDown(domState) {
+		_, err = l.setDomainXML(vm, wantedSpec)
```
do we need to update the dom object here if we're redefining the domain?

rmohr replied:
good question, will check that ...

rmohr commented:
retest this please

davidvossel commented on pkg/virt-controller/services/template.go:
```diff
@@ -65,7 +64,7 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VM) (*kubev1.Pod, error) {
 			{
 				Name:      "sockets",
-				MountPath: socketDir,
+				MountPath: t.socketBaseDir,
```
@rmohr 
I don't feel good about this change.  Previously each virt-launcher pod was isolated to it's own directory. With this change, all sock files are in a single common base directory that every virt-launcher pod has access to. 

Can we go back to the way this was?

rmohr replied:
Done, However, we now leave empy directories behind. I can't delete them reliably while there might be a chance that the virt-launcher pod is still running. Any ideas?

rmohr replied:
How concerned are you about this in general? At the end, we will have to start the virt-launcher pod via a privileged user to allow hostMounts, we control the image registry, and we can enter the virt-launcher namespace in a subsequent patch to also check if the virt-launcher contains a shared k8s secret between virt-launcher and virt-handler.

davidvossel replied:
I created an issue #421 for us to follow up on this.  I don't want to block this patch for that discussion.

In general, I'd like to restrict access for each component to only what it needs to see. That limits the possibilities to some degree.
